### PR TITLE
[GFX-560] Support for Mac Catalyst

### PR DESCRIPTION
**Pull request by palver123**
I had to add 2 compiler flags to clang as described in [our Confluence doc](https://shapr3d.atlassian.net/wiki/spaces/~112893953/pages/1221787924/Mac+Catalyst#New-platform) and then the code compiled for Mac Catalyst too. I had to adjust some compiler macros to make it work (Catalyst builds are treated as iOS builds and `IOS` macro is defined in the code).

I also had to add a new platform in the build script: `build.sh`

`./build.sh -gli -p ios,catalyst release` should build and install all libs:
 - arm64 iOS (for iPad) and x86_64 iOS for iOS Simulator in `out/ios-release` directory
 - arm64 Catalyst (for M1 Macs) and x86_64 Catalyst for Intel Macs in `out/catalyst-release` directory

### CODE REVIEW

File: filament/backend/src/metal/MetalEnums.h [hunk 1]
```diff
@@ -128,22 +128,24 @@ constexpr inline MTLVertexFormat getMetalFormat(ElementType type, bool normalize
...
-        case TextureFormat::SRGB8_ALPHA8_ASTC_8x6: return MTLPixelFormatASTC_8x6_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_8x8: return MTLPixelFormatASTC_8x8_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_10x5: return MTLPixelFormatASTC_10x5_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_10x6: return MTLPixelFormatASTC_10x6_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_10x8: return MTLPixelFormatASTC_10x8_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_10x10: return MTLPixelFormatASTC_10x10_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_12x10: return MTLPixelFormatASTC_12x10_sRGB;
-        case TextureFormat::SRGB8_ALPHA8_ASTC_12x12: return MTLPixelFormatASTC_12x12_sRGB;
-        default: break;
+    if (@available(iOS 8, macCatalyst 14, *)) {
```
Comment (veghbalazs) [inline filament/backend/src/metal/MetalEnums.h:131]: If I'm not mistaken, this function is intended to convert between enums AND check for support at the same time. Note that not all GPUs running Catalyst on macOS 11 support ASTC compression. Again, don't forget about "normal" macOS, where ASTC is supported by Apple GPUs. (Same goes for EAC/ETC2 below).

File: filament/backend/src/metal/MetalEnums.h [hunk 2]
```diff
@@ -186,6 +188,25 @@ inline MTLPixelFormat getMetalFormat(TextureFormat format) noexcept {
     }
 #endif
 
+#if defined(IOS)
+    if (@available(iOS 8, macCatalyst 14, *)) {
+        switch (format) {
+            // EAC / ETC2 formats are only available on iPhone.
```
Comment (veghbalazs) [inline filament/backend/src/metal/MetalEnums.h:194]: Outdated comment, should be saying Apple GPUs.

File: third_party/clang/iOS.cmake [hunk 3]
```diff
@@ -104,6 +108,9 @@ IF (NOT DEFINED IOS_MIN_TARGET)
 ENDIF()
 
 SET(IOS_COMMON_FLAGS "-m${PLATFORM_FLAG_NAME}-version-min=${IOS_MIN_TARGET}")
+if (CATALYST)
+  SET(IOS_COMMON_FLAGS "${IOS_COMMON_FLAGS} -target x86_64-apple-ios13.0-macabi")
```
Comment (veghbalazs) [inline third_party/clang/iOS.cmake:112]: What if I wish to build for Catalyst with version 14.0 (Big Sur) as minimum supported version? Shouldn't it be like this?
```suggestion
  SET(IOS_COMMON_FLAGS "${IOS_COMMON_FLAGS} -target x86_64-apple-ios${IOS_MIN_TARGET}-macabi")
```

**Review by veghbalazs — COMMENTED** [review]
I reviewed with a somewhat broader scope in mind, than what the introduction of Mac Catalyst support would suggest. With the arrival of Apple GPUs to the Mac (including Catalyst), a whole lot of device capabilities - previously exclusive to iOS - became a possibility on macOS too. 

The introduction of Mac Catalyst and Apple GPUs on Mac break the same assumptions about GPU capabilities, therefore I think it makes sense to revise these together.

The function `getMetalFormat(TextureFormat)` seems to need the most thorough rewrite, other changes look fairly easy to handle.

File: third_party/clang/iOS.cmake (hunk 3 above)
Comment (palver123) [reply third_party/clang/iOS.cmake:112]: Have you found any documentation on `x86_64-apple-ios14.0-macabi`? Does it even exist?

File: filament/backend/src/metal/MetalEnums.h (hunk 2 above)
Comment (palver123) [reply filament/backend/src/metal/MetalEnums.h]: 5c2c3efc2f8e3509c0c168d3e56a1c3f27386fb9

File: third_party/clang/iOS.cmake (hunk 3 above)
Comment (palver123) [reply third_party/clang/iOS.cmake]: You are right, see 27573cb3b5629bb1989a786612ee500f84984b79

File: filament/backend/src/metal/MetalHandles.mm [hunk 4]
```diff
@@ -857,7 +857,11 @@ static MTLPixelFormat decidePixelFormat(id<MTLDevice> device, TextureFormat form
     descriptor.sampleCount = samples;
     descriptor.usage = MTLTextureUsageRenderTarget;
 #if defined(IOS)
-    descriptor.resourceOptions = MTLResourceStorageModeMemoryless;
+    if (@available(macCatalyst 14.0, *)) {
+        descriptor.resourceOptions = MTLResourceStorageModeMemoryless;
```
[the remark this answers is not in the document]
Comment (palver123) [reply filament/backend/src/metal/MetalHandles.mm]: Hopefully covered all cases in 76c20e123dcd370310cf7e268419dfc9b79ef7fd

File: filament/backend/src/metal/MetalDriver.mm [hunk 5]
```diff
@@ -662,11 +662,11 @@
 }
 
 bool MetalDriver::isFrameBufferFetchSupported() {
-#if defined(IOS) && !defined(FILAMENT_IOS_SIMULATOR)
-    return true;
-#else
+    // Newer Intel cards support it too, but it's not documented
+    if (@available(macOS 10.15, iOS 13, *)) {
+        return [mContext->device supportsFamily:MTLGPUFamilyApple1];
```
[the remark this answers is not in the document]
Comment (palver123) [reply filament/backend/src/metal/MetalDriver.mm:673]: 76c20e123dcd370310cf7e268419dfc9b79ef7fd

File: build.sh [hunk 6]
```diff
@@ -256,6 +258,19 @@ function build_desktop {
     fi
 }
 
+function build_desktop_tools_for_ios {
+    # Supress intermediate desktop tools install and Java
```
Comment (marton-shapr3d) [inline build.sh:262]: nit^99: suppress

File: filament/backend/src/metal/MetalDriver.mm [hunk 7]
```diff
@@ -1008,11 +1015,6 @@
                                                                width:(srcTexture.width >> miplevel)
                                                               height:(srcTexture.height >> miplevel)
                                                            mipmapped:NO];
-#if defined(IOS)
```
Comment (marton-shapr3d) [inline filament/backend/src/metal/MetalDriver.mm:1011]: why was this removed?

File: filament/backend/src/metal/MetalEnums.h [hunk 8]
```diff
@@ -25,6 +25,32 @@
...
+        EAC_ETC,
+        ASTC
+    };
+
+    bool IsSupported(id<MTLDevice> device, TextureCompressionFamily family) {
+        assert(device);
+#if defined(FILAMENT_IOS_SIMULATOR)
+        return false;
+#elif defined(IOS) && !TARGET_OS_MACCATALYST
+        return true;
```
Comment (marton-shapr3d) [inline filament/backend/src/metal/MetalEnums.h:39]: this is practically correct for Shapr3D supported devices, but would not be applicable for Filament on its own (e.g. upstreaming), as not all iOS devices support ASTC

File: filament/backend/src/metal/MetalEnums.h [hunk 9]
```diff
@@ -126,64 +152,100 @@ constexpr inline MTLVertexFormat getMetalFormat(ElementType type, bool normalize
...
+                case TextureFormat::SRGB8_ALPHA8_ASTC_10x5: return MTLPixelFormatASTC_10x5_sRGB;
+                case TextureFormat::SRGB8_ALPHA8_ASTC_10x6: return MTLPixelFormatASTC_10x6_sRGB;
+                case TextureFormat::SRGB8_ALPHA8_ASTC_10x8: return MTLPixelFormatASTC_10x8_sRGB;
+                case TextureFormat::SRGB8_ALPHA8_ASTC_10x10: return MTLPixelFormatASTC_10x10_sRGB;
+                case TextureFormat::SRGB8_ALPHA8_ASTC_12x10: return MTLPixelFormatASTC_12x10_sRGB;
+                case TextureFormat::SRGB8_ALPHA8_ASTC_12x12: return MTLPixelFormatASTC_12x12_sRGB;
+                default: break;
+            }
+
+            // Only iOS 13.0 and above supports the ASTC HDR profile. Older versions of iOS fallback to LDR.
```
Comment (marton-shapr3d) [inline filament/backend/src/metal/MetalEnums.h:176]: I'm a bit confused here, as ASTC HDR is GPU families Apple6/7, and not just depending on iOS version

File: build.sh (hunk 6 above)
Comment (palver123) [reply build.sh:262]: Yes I noticed it too, and decided to leave it as it is

File: filament/backend/src/metal/MetalDriver.mm (hunk 7 above)
Comment (palver123) [reply filament/backend/src/metal/MetalDriver.mm:1011]: Because the default value is `Shared` on iOS and `Managed` on Catalyst and native macOS. The intent of the original author was to use `Shared` on iOS and `Managed` on native macOS. So instead of adding more macros to handle Catalyst I just removed the assignment altogether and let the default values persists. This is valid, because the descriptor itself was created a few lines above and the `storageMode` property is still on its default value.

File: filament/backend/src/metal/MetalEnums.h (hunk 8 above)
Comment (palver123) [reply filament/backend/src/metal/MetalEnums.h]: We refactored it with @veghbalazs to handle old devices: eedca31feddd656d676ff4fb7074527d20d6a612

File: filament/backend/src/metal/MetalEnums.h (hunk 9 above)
Comment (palver123) [reply filament/backend/src/metal/MetalEnums.h]: You are right, I missed the relevant section in [Metal Feature Set Tables, page 8](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf). Fixed in eedca31feddd656d676ff4fb7074527d20d6a612

File: filament/backend/src/metal/MetalDriver.mm (hunk 7 above)
Comment (marton-shapr3d) [reply filament/backend/src/metal/MetalDriver.mm:1011]: thanks

**Review by marton-shapr3d — APPROVED** [review]
LGTM, but I only checked the code and didn't build. please make sure another reviewer does it

File: CMakeLists.txt [hunk 10]
```diff
@@ -25,6 +25,8 @@ option(FILAMENT_SUPPORTS_XCB "Include XCB support in Linux builds" ON)
 
 option(FILAMENT_SUPPORTS_XLIB "Include XLIB support in Linux builds" ON)
 
+option(CATALYST "Build for Mac Catalyst in iOS builds" OFF)
```
Comment (veghbalazs) [inline CMakeLists.txt:28]: Not needed after the changes above.

File: build.sh [hunk 11]
```diff
@@ -589,34 +614,30 @@ function build_ios_target {
...
+    local archive_filename="filament-${archive_suffix}.tgz"
 
-    if [[ -d "out/ios-${lc_target}/filament" ]]; then
+    if [[ -d "${IOS_INSTALLED_LIBS_DIR}" ]]; then
         if [[ "${ISSUE_ARCHIVES}" == "true" ]]; then
-            echo "Generating out/filament-${lc_target}-ios.tgz..."
-            cd "out/ios-${lc_target}"
-            tar -czvf "../filament-${lc_target}-ios.tgz" filament
+            echo "Generating out/${archive_filename}..."
+            cd "${IOS_INSTALLED_LIBS_DIR}../../"
```
Comment (veghbalazs) [inline build.sh:632]: So the output of `archive_ios` now depends on the parameters `build_ios_target` was last invoked with? This looks like a fragile dependency,  can't we jump to the folder directly like this? 
```sh
cd "out/${archive_suffix}-${lc_target}"
```
(Though after this, the name `archive_suffix` might need some rethinking, and the first parameter "Debug" or "Release" has to be brought back).

File: third_party/clang/iOS.cmake [hunk 12]
```diff
@@ -104,6 +108,9 @@ IF (NOT DEFINED IOS_MIN_TARGET)
 ENDIF()
 
 SET(IOS_COMMON_FLAGS "-m${PLATFORM_FLAG_NAME}-version-min=${IOS_MIN_TARGET}")
+if (CATALYST)
+  SET(IOS_COMMON_FLAGS "${IOS_COMMON_FLAGS} -target ${IOS_ARCH}-apple-ios${IOS_MIN_TARGET}-macabi")
+endif()
```
Comment (veghbalazs) [inline third_party/clang/iOS.cmake:113]: At first glance I wasn't too happy about patching this toolchain file, which comes from [opensource.apple.com](https://opensource.apple.com/source/clang/clang-800.0.42.1/src/cmake/platforms/iOS.cmake.auto.html), but I see this has been the practice before upstream. It's definitely not our concern right now (or at all), but https://github.com/leetal/ios-cmake could be a useful replacement, with built-in support for Universal libs, Catalyst and Apple Silicon.

File: build.sh (hunk 11 above)
Comment (palver123) [reply build.sh]: 74fd2bd3fffbecb601faa377ae5c761f5d0b8530 (I realized that I accidentally removed `debug`/`release` from the archive filename, so thanks for pointing ot)

File: CMakeLists.txt (hunk 10 above)
Comment (palver123) [reply CMakeLists.txt]: 74fd2bd3fffbecb601faa377ae5c761f5d0b8530

File: third_party/clang/iOS.cmake [hunk 13]
```diff
@@ -22,6 +22,10 @@ if(PLATFORM_NAME STREQUAL "iphonesimulator")
     # The simulator only supports iOS >= 13.0
     set(IOS_MIN_TARGET "13.0")
 endif()
+if (CATALYST)
```
[the remark this answers is not in the document]
Comment (palver123) [reply third_party/clang/iOS.cmake]: 74fd2bd3fffbecb601faa377ae5c761f5d0b8530

File: build.sh [hunk 14]
```diff
@@ -32,7 +32,7 @@ function print_help {
     echo "    -m"
     echo "        Compile with make instead of ninja."
     echo "    -p platform1,platform2,..."
-    echo "        Where platformN is [desktop|android|ios|webgl|all]."
+    echo "        Where platformN is [desktop|android|ios|catalyst|webgl|all]."
     echo "        Platform(s) to build, defaults to desktop."
     echo "        Building for iOS will automatically perform a partial desktop build."
```
[the remark this answers is not in the document]
Comment (palver123) [reply build.sh]: Added WebGL and Catalyst in e939ad4b6758e89955eb36ac44df97d0bfb16379

File: build.sh [hunk 15]
```diff
@@ -32,7 +32,7 @@ function print_help {
     echo "    -m"
     echo "        Compile with make instead of ninja."
     echo "    -p platform1,platform2,..."
-    echo "        Where platformN is [desktop|android|ios|webgl|all]."
+    echo "        Where platformN is [desktop|android|ios|catalyst|webgl|all]."
```
[the remark this answers is not in the document]
Comment (palver123) [reply build.sh:35]: Well yes, but you have to be careful to not exclude OpenGL from iOS if called like this: `build.sh -p ios,catalyst release`
e939ad4b6758e89955eb36ac44df97d0bfb16379

File: build.sh [hunk 16]
```diff
@@ -647,14 +667,50 @@ function build_ios {
 
         if [[ "${BUILD_UNIVERSAL_LIBRARIES}" == "true" ]]; then
             build/ios/create-universal-libs.sh \
-                -o out/ios-release/filament/lib/universal \
-                out/ios-release/filament/lib/arm64 \
-                out/ios-release/filament/lib/x86_64
-            rm -rf out/ios-release/filament/lib/arm64
-            rm -rf out/ios-release/filament/lib/x86_64
+                -o "${IOS_INSTALLED_LIBS_DIR}/universal" \
```
[the remark this answers is not in the document]
Comment (palver123) [reply build.sh:671]: AFAIK, in Bash, you have two options for return values: global variable or echoing to stdout. Neither is clean. You can also use [dynamic scoping](https://riptutorial.com/bash/example/8094/dynamic-scoping-in-action) to communicate between caller and callee. I see several global variables in this script so I think it is perfectly fine to add 1 more. 
IMHO`IOS_INSTALLED_LIBS_DIR` is not worse than the code duplication was, assuming the install directory path everywhere.
